pkg/operator/awstagcontroller.go

```diff
@@ -0,0 +1,314 @@
+	if cr.Spec.Storage.S3 == nil {
+		return nil
+	}
+
+	// make a copy to avoid changing the cached data
+	cr = cr.DeepCopy()
+
+	infra, err := c.infraConfigClient.Get(
+		context.Background(),
+		defaults.InfrastructureResourceName,
+		metav1.GetOptions{},
+	)
+	if err != nil {
+		klog.Errorf("failed to fetch Infrastructure resource: %v", err)
+		return err
+	}
+
+	// Filtering tags based on validation
+	infraTagSet := filterPlatformStatusTags(infra)
+	klog.V(5).Infof("tags read from Infrastructure resource: %v", infraTagSet)
+
+	// Create a driver with the current configuration
+	ctx := context.Background()
+	driver := s3.NewDriver(ctx, cr.Spec.Storage.S3, &c.listers.StorageListers, c.featureGateAccessor)
+
+	s3TagSet, err := driver.GetStorageTags()
+	if err != nil {
+		klog.Errorf("failed to fetch storage tags: %v", err)
+		return err
+	}
+	klog.Infof("tags read from storage resource: %v", s3TagSet)
+
+	tagUpdatedCount := syncInfraTags(s3TagSet, infraTagSet)
+	if tagUpdatedCount > 0 {
+		if err := driver.PutStorageTags(s3TagSet); err != nil {
+			klog.Errorf("failed to update/append tagset of %s s3 bucket: %v", driver.ID(), err)
+			c.event.Warningf("UpdateAWSTags",
+				"Failed to update/append tagset of %s s3 bucket", driver.ID())
+		}
+		klog.Infof("successfully updated/appended %d tags, tagset: %+v", tagUpdatedCount, s3TagSet)
+		c.event.Eventf("UpdateAWSTags",
+			"Successfully updated tagset of %s s3 bucket", driver.ID())
+	}
+
+	return nil
+}
+
+// filterPlatformStatusTags is for reading and filter user tags present in
+// Platform Status of Infrastructure config.
+func filterPlatformStatusTags(infra *configv1.Infrastructure) map[string]string {
+	infraTagSet := map[string]string{}
+	for _, statusTags := range infra.Status.PlatformStatus.AWS.ResourceTags {
+		if err := validateUserTag(statusTags.Key, statusTags.Value); err != nil {
+			klog.Warningf("validation failed for tag(%s:%s): %v", statusTags.Key, statusTags.Value, err)
+			continue
+		}
+		infraTagSet[statusTags.Key] = statusTags.Value
+	}
+	return infraTagSet
+}
+
+// syncInfraTags synchronizes the tags obtained from S3 bucket and Infrastructure CR.
+// this modifies the s3TagSet based on new tags which are added and update the value to a key if it has changed.
+func syncInfraTags(s3TagSet map[string]string, infraTagSet map[string]string) int {
+	tagUpdatedCount := 0
+	for key, value := range infraTagSet {
+		val, ok := s3TagSet[key]
+		if !ok || val != value {
+			klog.V(5).Infof("%s tag will be added/updated with value %s", key, value)
+			s3TagSet[key] = value
+			tagUpdatedCount++
+		}
+	}
+	return tagUpdatedCount
+}
+
+// validateUserTag is for validating the user defined tags in Infrastructure CR
+func validateUserTag(key, value string) error {
+	if !tagKeyRegex.MatchString(key) {
+		return fmt.Errorf("key has invalid characters or length")
+	}
+	if strings.EqualFold(key, "Name") {
+		return fmt.Errorf("key cannot be customized by user")
```

Comment:
Let's make sure this error includes the key, just like the installer one does: https://github.com/openshift/installer/blob/master/pkg/types/aws/validation/platform.go#L78 (thanks for the link btw). We could even use the exact error string.
i.e 
```suggestion
		return fmt.Errorf("Name key is not allowed for user defined tags")
```